streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -1236,23 +1236,34 @@ object ZSink extends ZSinkPlatformSpecificConstructors with Serializable {
   /**
    * Creates a sink accumulating incoming values into a map.
    * Key of each element is determined by supplied function.
+   * Combines elements with same key with supplied function f.
+   *
    */
-  final def collectAllToMap[K, A](key: A => K): ZSink[Any, Nothing, Nothing, A, Map[K, A]] =
-    foldLeft[A, Map[K, A]](Map.empty[K, A])((map, element) => map + (key(element) -> element))
+  final def collectAllToMap[K, A](key: A => K)(f: (A, A) => A): Sink[Nothing, Nothing, A, Map[K, A]] =
+    foldLeft[A, Map[K, A]](Map.empty)((curMap, a) => {
+      val k = key(a)
+      curMap.get(k).fold(curMap.updated(k, a))(v => curMap.updated(k, f(v, a)))
+    })
 
   /**
    * Creates a sink accumulating incoming values into a map of maximum size `n`.
    * Key of each element is determined by supplied function.
+   *
+   * Combines elements with same key with supplied function f.
    */
-  final def collectAllToMapN[K, A](n: Long)(key: A => K): Sink[Nothing, A, A, Map[K, A]] = {
-    type State = (Map[K, A], Boolean)
-    def f(state: State, a: A): (State, Chunk[A]) = {
-      val newMap = state._1 + (key(a) -> a)
-      if (newMap.size > n) ((state._1, false), Chunk.single(a))
-      else if (newMap.size == n) ((newMap, false), Chunk.empty)
-      else ((newMap, true), Chunk.empty)
+  final def collectAllToMapN[K, A](n: Long)(key: A => K)(f: (A, A) => A): Sink[Nothing, A, A, Map[K, A]] = {
+
+    def inner(curMap: Map[K, A], a: A): (Map[K, A], Chunk[A]) = {
+      val k = key(a)
+      curMap
+        .get(k)
+        .fold(
+          if (curMap.size >= n) (curMap, Chunk.single(a))
```

Comment:
I think you need to signal that you don't want to continue here in the same way the previous implementation did. Otherwise, this probably won't work right with `ZStream#aggregate`.

Best add a test or two for verifying that behavior.